platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultResolvedDependency.java

```diff
@@ -48,15 +48,22 @@ public class DefaultResolvedDependency implements ResolvedDependency, Dependency
     private final ListMultimap<ResolvedDependency, ResolvedArtifactSet> parentArtifacts = ArrayListMultimap.create();
     private final String variantName;
     private final ModuleVersionIdentifier moduleVersionId;
-    private final BuildOperationExecutor buildOperationProcessor;
+    private final BuildOperationExecutor buildOperationExecutor;
+    private final ResolutionHost resolutionHost;
     private final Set<ResolvedArtifactSet> moduleArtifacts;
     private final Map<ResolvedDependency, Set<ResolvedArtifact>> allArtifactsCache = new HashMap<>();
     private Set<ResolvedArtifact> allModuleArtifactsCache;
 
-    public DefaultResolvedDependency(String variantName, ModuleVersionIdentifier moduleVersionId, BuildOperationExecutor buildOperationProcessor) {
+    public DefaultResolvedDependency(
+        String variantName,
+        ModuleVersionIdentifier moduleVersionId,
+        BuildOperationExecutor buildOperationExecutor,
+        ResolutionHost resolutionHost
```

Comment:
🤔 Having to supply a resolution host to create a resolved dependency seems wrong.  The resolution is *finished* now, we just have the dependency that was resolved.  Why are we keeping a "backreference" to what resolved it?

```
The "Host" or owner of a resolution -- the thing in charge of the resolution, or the thing being resolved.

The purpose of this type is to be a configuration-cache compatible representation of the thing being resolved. This type should remain as minimal as possible.
``` 

This interface seems to be more than that with the failure handling methods.  It's certainly not "as minimal as possible". 

Could this be split into 2 types: `ResolutionHost` and `ResolutionFailureMapper` or something?

Reply:
`ResolvedDependency` is a legacy type, tied to `ResolvedConfiguration` and `LenientConfiguration`. As we can see here, resolution is _not_ completed at this point. We are indeed resolving artifacts in this method here.

Reply:
Yeah, that's fair.  Then it's the name of this type itself that's the source of confusion.  It should be `DefaultResolvableDependency` or something.

When is it planned to remove this?  Is there already a ticket?  If so, I'm not concerned and can Approve.

Reply:
I agree the way `ResolutionHost` works ATM is not ideal. We should not tie exception-rethrowing to the thing owning the resolution.

However, I think we should get rid of `rethrowFailure` entirely and just throw exceptions outright. But I think that is a separate issue

Reply:
We will remove this once we remove `ResolvedConfiguration` and `LenientConfiguration`. I'm not sure we have an issue for that atm but we've been wanting to remove it for a long time. 

I do have a PR out currently that works to notify the community that these types are in maintenance mode. If you feel strongly about these, feel free to review

https://github.com/gradle/gradle/pull/30316

Reply:
That PR seems to be about different types (it doesn't mention `ResolvedDependency`) and also isn't about work on `ResolutionHost`.

Any time we decide to defer work until later, creating a new issue (even an `a:chore` issue for minor things) is a good idea, because it lets us all reference it, link to it, comment on it, and associate it with this PR, so that other readers can find the future plans.

Reply:
`ResolvedDependency` is only accessible through `ResolvedConfiguration` and `LenientConfiguration`

Reply:
I'm not sure we should write an issue at the moment. There are blockers to deprecating the above types, so the issue is not actionable at the moment. The linked PR is the most actionable things we can do right now

Reply:
Though I can write an issue that tracks the blockers to deprecating these types if you think thats useful

Reply:
I do think it's worth adding mention this type to the existing issue so that it can be easily found by searching, and creating a separate issue for addressing/splitting `ResolutionHost` (if you don't intend to do that any time soon).  Let's get this into the tracker before we forget about it.

Reply:
https://github.com/gradle/gradle/issues/30550